ssds/checksum.py

```diff
@@ -28,3 +28,44 @@ def compute_composite_etag(etags: typing.List[str]) -> str:
     bin_md5 = b"".join([binascii.unhexlify(etag) for etag in etags])
     composite_etag = md5(bin_md5).hexdigest() + "-" + str(len(etags))
     return composite_etag
+
+class UnorderedChecksum:
+    def __init__(self):
+        raise NotImplementedError()
+
+    def update(self, chunk_number: int, data: bytes):
+        raise NotImplementedError()
+
+    def hexdigest(self) -> str:
+        raise NotImplementedError()
+
+class S3EtagUnordered(UnorderedChecksum):
+    def __init__(self):
+        self._checksums: List[Tuple[int, str]] = list()
+
+    def update(self, chunk_number: int, data: bytes):
+        self._checksums.append((chunk_number, md5(data).hexdigest()))
+
+    def hexdigest(self) -> str:
+        return compute_composite_etag([cs[1] for cs in sorted(self._checksums)])
+
+class GScrc32cUnordered(UnorderedChecksum):
+    def __init__(self):
+        self._current_chunk_number = 0
+        self._chunks: Dict[int, bytes] = dict()
+        self._checksum = crc32c()
+
+    def update(self, chunk_number: int, data: bytes):
+        self._chunks[chunk_number] = data
+        for chunk_number in sorted(self._chunks.keys()):
+            if self._current_chunk_number == chunk_number:
+                self._checksum.update(self._chunks[chunk_number])
+                del self._chunks[chunk_number]
+                self._current_chunk_number += 1
+            else:
+                break
+
+    def hexdigest(self) -> str:
+        for chunk_number in sorted(self._chunks.keys()):
+            self._checksum.update(self._chunks[chunk_number])
```

Comment:
It's slightly confusing to me that the input for update is `self._chunks[chunk_number]` but that it looks like it should take two inputs (`update(self, chunk_number: int, data: bytes)`)?  One value of which it immediately reforms as another `self._chunks[chunk_number]` in `update()`?

Maybe I'm missing something.

Reply:
The internal checksum does not accept chunk number.

Reply:
Ah, you're calling two `update()` functions.  Do'h.